warp/yul/DeadcodeEliminator.py

```diff
@@ -0,0 +1,171 @@
+        [var_set, nodes] = list(map(list, zip(*lis)))
+        live_vars = reduce(lambda s1, s2: s1 | s2, var_set, set())
+        return (live_vars, nodes)
+
+    def visit_literal(self, node: ast.Literal) -> Tuple[LiveVars, ast.Literal]:
+        return (set(), node)
+
+    def visit_identifier(self, node: ast.Identifier) -> Tuple[LiveVars, ast.Identifier]:
+        return (set([node.name]), node)
+
+    def visit_typed_name(self, node: ast.TypedName):
+        return (set([node.name]), node)
+
+    def visit_function_call(
+        self, node: ast.FunctionCall
+    ) -> (LiveVars, ast.FunctionCall):
+        (live_vars, arguments) = self.visit(node.arguments)
+        return (
+            live_vars,
+            ast.FunctionCall(function_name=node.function_name, arguments=arguments),
+        )
+
+    def visit_expression_statement(
+        self, node: ast.ExpressionStatement
+    ) -> (LiveVars, ast.ExpressionStatement):
+        (live_vars, _) = self.visit(node.expression)
+        return (live_vars, node)
+
+    def visit_assignment(self, node: ast.Assignment) -> Tuple[LiveVars, ast.Assignment]:
+        (live_vars, _) = self.visit(node.value)
+        return (live_vars, node)
+
+    def visit_variable_declaration(
+        self, node: ast.VariableDeclaration
+    ) -> (LiveVars, ast.VariableDeclaration):
+        if node.value == None:
+            return (set(), node)
+        (live_vars, _) = self.visit(node.value)
+        return (live_vars, node)
+
+    def visit_function_definition(
+        self, node: ast.FunctionDefinition
+    ) -> (LiveVars, ast.FunctionDefinition):
+        # We need to ignore the generated getters from kudu since their
+        # body is only instantiated in ToCairoVisitor
+        if node.name.startswith("getter"):
+            return (set(), node)
+        (live_vars, body) = self.visit(
+            node.body, return_vars=set([n.name for n in node.return_variables])
+        )
+        live_vars = live_vars - set([arg.name for arg in node.parameters])
+        # The warp ifs are the only place where the if branch has a return in each branch
+        # if node.name.startswith("__warp_if"):
+        #     live_vars -= set([arg.name for arg in node.return_variables])
+        return (
+            live_vars,
+            ast.FunctionDefinition(
+                name=node.name,
+                parameters=node.parameters,
+                return_variables=node.return_variables,
+                body=body,
+            ),
+        )
+
+    def visit_if(
+        self, node: ast.If, live_vars: Set[str] = None, return_vars: Set[str] = None
+    ) -> Tuple[LiveVars, ast.If]:
+        if any(
+            [
+                isinstance(statement, ast.Leave)
+                for statement in node.body.statements
+                + (node.else_body.statements if node.else_body else ())
+            ]
+        ):
+            live_vars = return_vars | live_vars
```

Comment:
Just carping: `live_vars |= return_vars`

Reply:
This is not a good idea here, it replaces live_vars in place and can easily cause unexpected sets to be passed into functions

Reply:
Ah right, it's passed from somewhere else. 
In general, it seems that you never modify sets. Why not enforce that using `frozenset`? In that case `|=` is completely safe also.

`frozenset` might be bothersome to type and read, so we can shorten it by `fset = frozenset` somewhere above

Reply:
Thanks